internal/cmd/oohelperd/internal/nwcth/http.go

```diff
@@ -0,0 +1,110 @@
+package nwcth
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"net/http/cookiejar"
+	"net/url"
+	"strings"
+
+	"github.com/ooni/probe-cli/v3/internal/iox"
+	"github.com/ooni/probe-cli/v3/internal/runtimex"
+)
+
+// HTTPConfig configures the HTTP check.
+type HTTPConfig struct {
+	Jar       *cookiejar.Jar
+	Headers   map[string][]string
+	Transport http.RoundTripper
+	URL       *url.URL
+}
+
+// HTTPDo performs the HTTP check.
+func HTTPDo(ctx context.Context, config *HTTPConfig) (*CtrlHTTPRequest, *NextLocationInfo) {
+	req, err := newRequest(ctx, config.URL)
+	if err != nil {
+		return &CtrlHTTPRequest{Failure: newfailure(err)}, nil
+	}
+	// The original test helper failed with extra headers while here
+	// we're implementing (for now?) a more liberal approach.
+	for k, vs := range config.Headers {
+		switch strings.ToLower(k) {
+		case "user-agent":
+		case "accept":
+		case "accept-language":
+			for _, v := range vs {
+				req.Header.Add(k, v)
+			}
+		}
+	}
+	// redirectReq is a clone of the initial request, with possibly modified headers
+	var redirectReq *http.Request
+
+	jar := config.Jar
+	if jar == nil {
+		jar, err = cookiejar.New(nil)
+		runtimex.PanicOnError(err, "cookiejar.New failed")
+	}
+	client := http.Client{
+		CheckRedirect: func(r *http.Request, reqs []*http.Request) error {
+			redirectReq = r
```

Comment:
Are you sure `CheckRedirect` is called from the same goroutine in which we define `redirectReq`? Did you try running the code with `-race`? I am not sure we have enough guarantee for the docs to disregard the possibility that the `CheckRedirect` callback will run in another goroutine.

Reply:
Testing with `-race` says ok. I am wondering whether we even need to store this request. We conserve the cookie jar, so the cookies are taken care of. Other than that we are only passing along "user-agent", "accept", "accept-language" headers. I'm trying to understand whether they could be changed by redirects, but I don't think so.

Reply:
> I am wondering whether we even need to store this request

Good question. My instinct is that probably we don't.